Include field annotations for native queries too

snoe opened the pull request:
Persistence will replace a source-table source-query with a native
query, but preprocess has still filled in source-metadata with all of
the relevant field-ids expected to be returned. With this change we
include field info from the store in the same way that mbql-cols does.
This allows persisted models to honor field settings like `:visibility
:details-only`.

snoe committed:
Include field annotations for native queries too

Persistence will replace a source-table source-query with a native
query, but preprocess has still filled in source-metadata with all of
the relevant field-ids expected to be returned. With this change we
include field info from the store in the same way that mbql-cols does.
This allows persisted models to honor field settings like `:visibility
:details-only`.

dpsutton commented:
Little more background:
New json feature will "explode" the json keys and automatically hide the json field. So an object like `{"a": 1, "b": 2}` would be hidden but you'd have two columns "json -> a" and "json -> b". But behind the scenes this just works on setting the visibility of the underlying field.

This visibility is NOT part of the metadata, and cannot be. It has to be fetched from the field information in the db. (We don't want to update N questions when a field's visibility changes). But since we have the field id's laying around, we can fix it. There are two choices:
- special case these persistence substitutions or
- augment native query annotations. We generally have no hope to add extra information for native queries. But this is a special case, but not the only case. Here we have the original metadata from the mbql query which of course has field ids in it. So when getting native column information, if we have a field id, look it up in the db.

We took the latter approach as it might benefit models in general when people add metadata information. This stuff is very sensitive though so we're watching for test failures.

snoe committed:
Force type of merge-source-metadata-col to map

By doing the lookup to store/field at the top of the merge, the type of
annotations coming through was a FieldInstance. Tests, at least, were
unhappy about this and it's better not to change it.

snoe committed:
Merge remote-tracking branch 'origin/master' into persist-models-include-field-annotations

snoe committed:
Resolve fields for ids in source-metadata

Makes sure that the qp/store has all the available fields for
annotations.

snoe committed:
Recursively find source-metadata field-ids for annotations

dpsutton commented on src/metabase/query_processor/middleware/resolve_fields.clj:
```diff
@@ -16,7 +16,9 @@
   [query]
   (let [ids (into (set (mbql.u/match (:query query) [:field (id :guard integer?) _] id))
-                  (keep :id (get-in query [:query :source-metadata])))]
+                  (->> (mbql.u/match (:query query) {:source-metadata source-metadata} source-metadata)
```
You are in a transducing context from the `into` so you could `(into (set ...) (comp cat (keep :id)) (mbql.u/match ...))` and skip the intermediate collections from the concat and keep.

snoe committed:
Use transducer as per review
